mlflow/spark.py

```diff
@@ -208,33 +215,26 @@ def log_model(
         spark_model = PipelineModel([spark_model])
     run_id = mlflow.tracking.fluent._get_or_start_run().info.run_id
     run_root_artifact_uri = mlflow.get_artifact_uri()
+    if _should_use_mlflowdbfs(run_root_artifact_uri):
+        mlflowdbfs_path = _mlflowdbfs_path(run_id, artifact_path)
+        with databricks_utils.MlflowCredentialContext(
+            get_databricks_profile_uri_from_artifact_uri(run_root_artifact_uri)
+        ):
+            try:
+                spark_model.save(mlflowdbfs_path)
+                return
+            except Py4JError as e:
+                raise MlflowException("failed to save Spark model via mlflowdbfs") from e
+
     # If the artifact URI is a local filesystem path, defer to Model.log() to persist the model,
     # since Spark may not be able to write directly to the driver's filesystem. For example,
     # writing to `file:/uri` will write to the local filesystem from each executor, which will
-    # be incorrect on multi-node clusters - to avoid such issues we just use the Model.log() path
-    # here.
-    if is_local_uri(run_root_artifact_uri):
-        return Model.log(
-            artifact_path=artifact_path,
-            flavor=mlflow.spark,
-            spark_model=spark_model,
-            conda_env=conda_env,
-            code_paths=code_paths,
-            dfs_tmpdir=dfs_tmpdir,
-            sample_input=sample_input,
-            registered_model_name=registered_model_name,
-            signature=signature,
-            input_example=input_example,
-            await_registration_for=await_registration_for,
-            pip_requirements=pip_requirements,
-            extra_pip_requirements=extra_pip_requirements,
-        )
-    model_dir = os.path.join(run_root_artifact_uri, artifact_path)
-    # Try to write directly to the artifact repo via Spark. If this fails, defer to Model.log()
-    # to persist the model
-    try:
-        spark_model.save(posixpath.join(model_dir, _SPARK_MODEL_PATH_SUB))
-    except Py4JError:
+    # be incorrect on multi-node clusters.
+    # If the artifact URI is not a local filesystem path we attempt to write directly to the
+    # artifact repo via Spark. If this fails, we defer to Model.log().
+    elif is_local_uri(run_root_artifact_uri) or not _maybe_save_model(
+        spark_model, posixpath.join(run_root_artifact_uri, artifact_path, _SPARK_MODEL_PATH_SUB)
```

Comment:
joining a URI with a path may not always produce a valid URI if the URI contains a fragment or query parameters. Can we use urllib to append these to the path component? I realize the preexisting code didn't handle this properly.